### ISSUE TITLE
Potential Issue with Classes

### ISSUE THREAD

**Issue by LethiferousMoose** [pester/Pester#1243]
## 1. General summary of the issue
So I'm not 100% sure what's going on here, it only seems to happen in pester tests so I'm going to log it here in case anyone has any ideas. I have a pester test for a module, it has a configuration class that loads up a JSON file. I have the class instance saved off in a script level variable in the module. What I was attempting to do was replace the original configuration class with a test instance in the unit test to allow me to load up test files. I create and assign the test class by referencing the script variable while **InModuleScope**. 

Here is a rough example of the code (only showing relevant part):
Base class (used in many different modules, lets call the module it resides in **Classes**)
```powershell
class Configuration {
    [string] GetConfigFileLocation() {
        # Some logic that reads configuration using a provided $PSScriptRoot and hard-coded file name.
    }
}
```

In Module (lets call it **TestModule**):
```powershell
using Module Classes

[ModuleConfiguration] $config = [ModuleConfiguration]::new()

class ModuleConfiguration : Configuration  {
    # Calls parent constructor using $PSScriptRoot that the base class uses to load configurations
}
```

In Unit Test
```powershell
using Module TestModule

InModuleScope TestModule {
    class TestConfiguration : Configuration{
        [string] $TestingConfig;

        TestConfiguration() : base() {
        }

        hidden [string] GetConfigFileLocation() {
            return $this.TestingConfig
        }
    }

    $Script:config = [TestConfig]::new()
}
```

## 2. Describe Your Environment

```powershell
Pester version     : 4.6.0 C:\Program Files\PowerShell\Modules\Pester\4.6.0\Pester.psd1
PowerShell version : 6.1.0
OS version         : Microsoft Windows NT 10.0.16299.0
```

## 3. Expected Behavior

The assignment should work without errors.
## 4.Current Behavior

Assignment fails with error
```
Cannot convert the "TestConfiguration" value of type "TestConfiguration" to type "ModuleConfiguration"
```
## 5. Possible Solution

No idea, not sure what is causing it.

## 6. Context

I may have over complicated this, I may just up adding a setter on the real base class to handle this, but I still wanted to log and issue and see if anyone had any ideas why this happens. It seems to only occur within Pester tests, but it may be a PowerShell bug.

**Comment by nohwnd** [comment]
The error says `SampeClass` (without `l`) but your code says `SampleClass`. Have you copied or wrote the message? I am trying to find out if this is just a typo, or some session related problem.

Are you able to replicate this successfully in a new PowerShell process?

**Comment by LethiferousMoose** [comment]
Yes that was a typo, I was using code I had written and I swapped in different class names.

**Comment by LethiferousMoose** [comment]
Updated description. Some relevant details:
- The module needs to be "using" a second module
- Test must  remove and re-import the module into the testing session

For the second bullet point. I've been doing this in all my tests in order to refresh the session with module changes since they normally only get loaded once per session.
```
Get-Module TestModule | Remove-Module -Force
Import-Module $PSScriptRoot\..\TestModule\TestModule.psm1 -Force
```
Is there a better way to accomplish this?

**Comment by Stephanevg** [comment]
Hi @LethiferousMoose 

I have been writing quite a few modules with classes. And one thing that I learned trhough time, is that the 'using module' statement is really just a source of problems; So I try to avoid it at all costs.
I might ahve a solution for you, but It would imply you re structure your modules code a bit.

If I need to have access to the instance of class that is located in a module, I generally write a function wrapper for it, add it to the exported functions, and make it return my instance.

## Example

```powershell

# classes are always defined at top of module file

class SampleClass {
 # do stuff
}

Function Get-SampleStuffinstance{

  $config = [SampleClass]::new()
  Return $Config
}


```
To load your module, the only thing you need to do, as add an import statement (no using module statement is required).

This might not be the best solution, but at least like this you don't have the problem of the double loading anymore.

cheers

**Comment by fmichaleczek** [comment]
@Stephanevg
I publish on Gist a sample implementation where the module "ModuleTypePOC" adds your module classes to the global scope via private Reflection. 

[Module Type Gist](https://gist.github.com/fmichaleczek/e91e8965a0f18ff793b027a30ec4fab3)

It's maybe bad, but I don't have enhanced classes with async/runspace/parallel features, it works enough for me, for coding, testing or building a new POC module.

After editing a class, if you enter 'Import-Module -Force' then 'Import-ModuleType' (with your module), it will refresh your global sessionstate with new version.

It's the same behavior as dotnet assemblies. 

I suggest to read this article to understand the underlayers : [ImplementingAssembly: Find PowerShell Classes in Modules](https://info.sapien.com/index.php/scripting/scripting-classes/implementingassembly-find-powershell-classes-in-modules)

Could this code help you in your scenario ?

**Comment by nohwnd** [comment]
Thanks for the detailed writeup. This is a PowerShell class limitation rather than a Pester bug: PowerShell compiles each class into its own dynamic assembly, so when the class gets re-parsed in another scope (here via `InModuleScope` + `using module`) it ends up with a new type identity. The names match but the runtime types don't, which is why the cast fails — you can see it in the `<103789fe>` assembly prefix in the error. Pester can't unify those identities, only the engine can.

Since you've already found a working approach, I'll close this. If you can still reproduce it on current PowerShell without Pester in the mix, please raise it on PowerShell/PowerShell.
